packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp

```diff
@@ -265,10 +312,51 @@ namespace Amesos2 {
       srcMap = importer_->getSourceMap ();
     }
 
-    const multivec_t source_mv (srcMap, new_data, lda, num_vecs);
+    multivec_t redist_mv (srcMap, num_vecs);
+
+    if ( distribution != NON_CONTIGUOUS_AND_ROOTED ) {
+      // Do this if GIDs contiguous - existing functionality
+      // Redistribute the output (multi)vector.
+      const multivec_t source_mv (srcMap, new_data, lda, num_vecs);
+      mv_->doImport (source_mv, *importer_, Tpetra::REPLACE);
+    }
+    else {
+      typedef typename multivec_t::dual_view_type dual_view_type;
+      typedef typename dual_view_type::host_mirror_space host_execution_space;
+      redist_mv.template modify< host_execution_space > ();
+
+      if ( redist_mv.isConstantStride() ) {
+        auto contig_local_view_2d = redist_mv.template getLocalView<host_execution_space>();
+        for ( size_t j = 0; j < num_vecs; ++j) {
+          auto av_j = new_data(lda*j, lda);
+          for ( size_t i = 0; i < lda; ++i ) {
+            contig_local_view_2d(i,j) = av_j[i];
+          }
+        }
+      }
```

Comment:
@mhoemmen thanks for the suggested change for the non-constant stride case when copying from the MultiVector to the Teuchos::Array. 
If you get a chance to take a glance and notice anything I missed please let me know.